ci: Add GitHub workflow for running QA experiments

nenadmilosevic95 opened the pull request:
Closes: #800 

This PR introduces a GitHub workflow that enables us to run QA experiments directly from Git.

It supports two default configurations:
- `north-america`: Deploys 4 nodes across four regions in North America.
- `global`: Deploys 20 nodes across 10 regions worldwide.

All nodes are provisioned on DigitalOcean, and the experiment duration can be configured.
Metrics are available in real time via Grafana, and a summary of key performance metrics (latency, block time, throughput) is uploaded to an S3 bucket.
The performance graphs at https://informal.systems/malachite/dashboard are generated from these files.

---

### PR author checklist

#### For all contributors

- [ ] Reference a GitHub issue
- [ ] Ensure the PR title follows the [conventional commits][conv-commits] spec
- [ ] Add a release note in [`RELEASE_NOTES.md`](/RELEASE_NOTES.md) if the change warrants it
- [ ] Add an entry in [`BREAKING_CHANGES.md`](/BREAKING_CHANGES.md) if the change warrants it

#### For external contributors

- [ ] Maintainers of Malachite are [allowed to push changes to the branch][gh-edit-branch]

[conv-commits]: https://www.conventionalcommits.org/en/v1.0.0/#summary
[gh-edit-branch]: https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests

nenadmilosevic95 commented on qa/terraform/templates/commands.tmpl:
```diff
@@ -45,6 +45,9 @@
 #  _change_config
 #  _create_hostname_links
 #  _parse_multiple_hosts
+# Undocumented commands used for geting prometheus metrics in .csv format
```
```suggestion
# Undocumented commands used for getting prometheus metrics in .csv format
```

nenadmilosevic95 commented on .github/workflows/qa.yml:
```diff
@@ -0,0 +1,122 @@
+          - "10"
+          - "15"
+
+  push:
```
```suggestion
```

nenadmilosevic95 commented on .github/workflows/qa.yml:
```diff
@@ -0,0 +1,122 @@
+          - "5"
+          - "10"
+          - "15"
+
+  push:
+    branches:
+      - nenad/800-run-qa-from-git
```
```suggestion
```

greg-szabo commented on code/scripts/nohup.out:
I think this file is unneccessary. You can add it to a `.gitignore` file if it pops up regularly.

greg-szabo reviewed (approved):
Looks good. Nice new functions!

Please remove the nohup.out file before merging.

ancazamfir commented on .github/workflows/qa.yml:
```diff
@@ -0,0 +1,118 @@
+name: QA
+
+on:
+  workflow_dispatch:
```
Should we run this on release as well?

romac replied:
Or we can we do it manually before the release to catch any regressions?

ancazamfir reviewed (approved):
great stuff! looking forward to running this/ checking results

nenadmilosevic95 commented:
> Looks good. Nice new functions!
> 
> Please remove the nohup.out file before merging.

Removed the file and added it to `.gitignore`.
Thanks, @greg-szabo, for reviewing the PR — and once again, for all the great work you put into it!

romac commented on .github/workflows/qa.yml:
```diff
@@ -0,0 +1,118 @@
+  workflow_dispatch:
+    inputs:
+      setup:
+        description: "Deployment setup:"
```
```suggestion
        description: "Deployment setup"
```

romac commented on .github/workflows/qa.yml:
```diff
@@ -0,0 +1,118 @@
+
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v4
```
Will need to add the branch to checkout here via `gitHub.event.inputs.branch` in order to satisfy https://github.com/informalsystems/malachite/pull/1023/files#r2088306372

adizere commented:
Nice work Nenad!

The measurements are live in the Dashboard: https://informal.systems/malachite/dashboard
It would be ideal to run this weekly in my view & use it for catching performance regressions

We still need to add support for navigating the history of measurements (eg go back in the past) in the Dashboard. Aaron is helping with that.

Another possible extension for the future is to add a Setup C - 100 validators. But not needed yet.

romac reviewed (approved):
Awesome work! Thank you @greg-szabo and @nenadmilosevic95 🙏
